spec/controllers/bookings_spec.cr

```diff
@@ -255,6 +255,139 @@ describe Bookings do
     guest.email.should eq("jon@example.com")
   end
 
+  # add support for configurable booking limits on resources
+  it "#create and #update should respect booking limits" do
+    WebMock.stub(:post, "#{ENV["PLACE_URI"]}/auth/oauth/token")
+      .to_return(body: File.read("./spec/fixtures/tokens/placeos_token.json"))
+    WebMock.stub(:post, "#{ENV["PLACE_URI"]}/api/engine/v2/signal?channel=staff/booking/changed")
+      .to_return(body: "")
+    WebMock.stub(:post, "#{ENV["PLACE_URI"]}/api/engine/v2/signal?channel=staff/guest/attending")
+      .to_return(body: "")
+
+    # Set booking limit
+    tenant = Tenant.query.find! { domain == "toby.staff-api.dev" }
+    tenant.booking_limits = JSON.parse(%({"desk": 2}))
+    tenant.save!
+
+    starting = 5.minutes.from_now.to_unix
+    ending = 40.minutes.from_now.to_unix
+    different_starting = 45.minutes.from_now.to_unix
+    different_ending = 55.minutes.from_now.to_unix
+
+    first_booking = Context(Bookings, JSON::Any).response("POST", "#{BOOKINGS_BASE}/",
+      body: %({"asset_id":"first_desk","booking_start":#{starting},"booking_end":#{ending},"booking_type":"desk"}),
+      headers: Mock::Headers.office365_guest, &.create)[1].as_h
+    first_booking["asset_id"].should eq("first_desk")
+
+    second_booking = Context(Bookings, JSON::Any).response("POST", "#{BOOKINGS_BASE}/",
+      body: %({"asset_id":"second_desk","booking_start":#{starting},"booking_end":#{ending},"booking_type":"desk"}),
+      headers: Mock::Headers.office365_guest, &.create)[1].as_h
+    second_booking["asset_id"].should eq("second_desk")
+
+    # Fail to create booking due to limit
+    not_created = Context(Bookings, JSON::Any).response("POST", "#{BOOKINGS_BASE}/",
+      body: %({"asset_id":"third_desk","booking_start":#{starting},"booking_end":#{ending},"booking_type":"desk"}),
+      headers: Mock::Headers.office365_guest, &.create)[0]
+    not_created.should eq(409)
+
+    # Create third booking at a different time
+    third_booking = Context(Bookings, JSON::Any).response("POST", "#{BOOKINGS_BASE}/",
+      body: %({"asset_id":"third_desk","booking_start":#{different_starting},"booking_end":#{different_ending},"booking_type":"desk"}),
+      headers: Mock::Headers.office365_guest, &.create)[1].as_h
+    third_booking["asset_id"].should eq("third_desk")
```

Comment:
I would turn these into a function that accepts args for the differences in the body since this code is repeated in a few of these tests

Reply:
I have created a helper method: #http_create_booking